firedrake/slate/slate.py

```diff
@@ -0,0 +1,673 @@
+        the tensor."""
+        return self.children.ufl_domains()
+
+    def ufl_domain(self):
+        """This function returns a single domain of integration occuring
+        in the tensor.
+
+        The function will fail if multiple domains are found."""
+        return self.children.ufl_domain()
+
+    def subdomain_data(self):
+        """Returns a mapping on the tensor:
+        `{domain:{integral_type: subdomain_data}}`."""
+        return self.children.subdomain_data()
+
+    @property
+    def operands(self):
+        """Return the operand of the inverse operator."""
+        return (self.children, )
+
+    def __str__(self, prec=None):
+        """String representation of the inverse of a SLATE tensor."""
+        return "(%s).inv" % self.children
+
+    def __repr__(self):
+        """SLATE representation of the inverse of a tensor."""
+        return "Inverse(%s)" % self.children
+
+
+class Transpose(Tensor):
+    """An abstract SLATE class representing the tranpose of a tensor."""
+
+    def __init__(self, A):
+        """Constructor for the Transpose class.
+
+        :arg A: a SLATE tensor."""
+        self.children = A
+        super(Transpose, self).__init__()
+
+    def arguments(self):
+        """Returns a tuple of arguments associated with the tensor."""
+        return self.children.arguments()[::-1]
+
+    def coefficients(self):
+        """Returns a tuple of coefficients associated with the tensor."""
+        return self.children.coefficients()
+
+    def ufl_domains(self):
+        """Returns the integration domains of the integrals associated with
+        the tensor."""
+        return self.children.ufl_domains()
+
+    def ufl_domain(self):
+        """This function returns a single domain of integration occuring
+        in the tensor.
+
+        The function will fail if multiple domains are found."""
+        return self.children.ufl_domain()
+
+    def subdomain_data(self):
+        """Returns a mapping on the tensor:
+        `{domain:{integral_type: subdomain_data}}`."""
+        return self.children.subdomain_data()
+
+    @property
+    def operands(self):
+        """Returns the operand of the transpose operator."""
+        return (self.children, )
+
+    def __str__(self, prec=None):
+        """String representation of a transposed tensor."""
+        return "(%s).T" % self.children
+
+    def __repr__(self):
+        """SLATE representation of a transposed tensor."""
+        return "Transpose(%s)" % self.children
+
+
+class UnaryOp(Tensor):
+    """An abstract SLATE class for representing unary operations on a
+    Tensor object."""
+
+    def __init__(self, A):
+        """Constructor for the UnaryOp class.
+
+        :arg tensor: a SLATE tensor."""
+
+        if not isinstance(A, Tensor):
+            raise Exception("Expecting a `slate.Tensor` object, not %r" % A)
+        self.children = A
+        super(UnaryOp, self).__init__()
```

Comment:
If you can, please call the super class `__init__` before assigning any attributes of `self`, so that I can be sure that you aren't using this ordering to facilitate hacks. This also applies to all the other places with this pattern.